File: compiler/src/main/java/com/raizlabs/android/dbflow/processor/definition/TypeConverterDefinition.java
package com.raizlabs.android.dbflow.processor.definition;

import com.raizlabs.android.dbflow.processor.Classes;
import com.raizlabs.android.dbflow.processor.model.ProcessorManager;

import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;

import java.util.List;

/**
 * Author: andrewgrosner
 * Contributors: { }
 * Description:
 */
public class TypeConverterDefinition {

    private TypeElement classElement;

    private TypeElement modelElement;

    private TypeElement dbElement;

    public TypeConverterDefinition(TypeElement classElement, ProcessorManager manager) {

        this.classElement = classElement;

        Types types = manager.getTypeUtils();

        DeclaredType typeConverterSuper = null;
        DeclaredType typeConverter = manager.getTypeUtils().getDeclaredType(manager.getElements().getTypeElement(Classes.TYPE_CONVERTER),
                types.getWildcardType(null, null), types.getWildcardType(null, null));

        for(TypeMirror superType: types.directSupertypes(classElement.asType())) {
            if(types.isAssignable(superType, typeConverter)) {
                typeConverterSuper = (DeclaredType) superType;
            }
        }

        if(typeConverterSuper != null) {
            List<? extends TypeMirror> typeArgs = typeConverterSuper.getTypeArguments();
            dbElement = manager.getElements().getTypeElement(typeArgs.get(0).toString());
            modelElement = manager.getElements().getTypeElement(typeArgs.get(1).toString());
        }
    }

    public TypeElement getModelElement() {
        return modelElement;
    }

    public TypeElement getDbElement() {
        return dbElement;
    }

    public TypeElement getClassElement() {
        return classElement;
    }

    public String getModelClassQualifiedName() {
        return modelElement.getQualifiedName().toString();
    }

    public String getQualifiedName() {
        return classElement.getQualifiedName().toString();
    }
}

File: processor/src/main/kotlin/com/dbflow5/processor/definition/TypeConverterDefinition.kt
package com.dbflow5.processor.definition

import com.dbflow5.annotation.TypeConverter
import com.dbflow5.processor.ClassNames
import com.dbflow5.processor.ProcessorManager
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.TypeName
import javax.lang.model.type.DeclaredType
import javax.lang.model.type.MirroredTypesException
import javax.lang.model.type.TypeMirror

/**
 * Description: Holds data about type converters in order to write them.
 */
class TypeConverterDefinition(
    typeConverter: TypeConverter?,
    val className: ClassName,
    typeMirror: TypeMirror, manager: ProcessorManager,
    val isDefaultConverter: Boolean) {

    val modelTypeName: TypeName?
    val dbTypeName: TypeName?
    val allowedSubTypes: List<TypeName>

    init {
        val allowedSubTypes: MutableList<TypeName> = mutableListOf()
        typeConverter?.let { converter ->
            try {
                converter.allowedSubtypes
            } catch (e: MirroredTypesException) {
                val types = e.typeMirrors
                types.forEach { allowedSubTypes.add(TypeName.get(it)) }
            }
        }
        this.allowedSubTypes = allowedSubTypes

        val types = manager.typeUtils

        var typeConverterSuper: DeclaredType? = null
        val typeConverterType = manager.typeUtils.getDeclaredType(manager.elements
            .getTypeElement(ClassNames.TYPE_CONVERTER.toString()))

        for (superType in types.directSupertypes(typeMirror)) {
            val erasure = types.erasure(superType)
            if (types.isAssignable(erasure, typeConverterType) || erasure.toString() == typeConverterType.toString()) {
                typeConverterSuper = superType as DeclaredType
            }
        }

        if (typeConverterSuper != null) {
            val typeArgs = typeConverterSuper.typeArguments
            dbTypeName = ClassName.get(typeArgs[0])
            modelTypeName = ClassName.get(typeArgs[1])
        } else {
            dbTypeName = null
            modelTypeName = null
        }
    }

}
